### ISSUE TITLE
Federation 

### ISSUE THREAD

**Issue by dessalines** [LemmyNet/lemmy#145]
Required items before we can enable ActivityPub in production. This list is probably incomplete, and we will add additional items once we get more familiar with the protocol. Our primary goal is to federate among Lemmy instances. Federation with other implementations (eg Mastodon or Pleroma) will come after that.

- [ ] Community Federation
  - [ ] Community actors (#369)
  - [ ] View remote communities ([!7](https://yerbamate.dev/dessalines/lemmy/pulls/7))
  - [ ] Outbox with list of posts created by the community
  - [ ] Also federate image posts
  - [ ] Federate comments
  - [ ] Inbox where new posts can be submitted
  - [ ] Remote upvoting and downvoting
  - [ ] Front page that includes remote posts
- [ ] User Federation
  - [ ] User actors (#369)
  - [ ] View remote user profiles
  - [ ] Users can follow remote communities (#150)
  - [ ] Outbox with list of posts created by the user, community is changed to announce (boost) these posts
  - [ ] Federated private messaging
- [ ] Optimisation
  - [ ] Store data from remote actors in local database (including inbox/outbox url, public key etc)
  - [ ] Store remote posts locally, possibly including images
  - [ ] Fully convert all lemmy db views to activitypub vocab
- [ ] Security and Error Handling
  - [x] webfinger (#149)
  - [ ] http signatures ([explained here](https://blog.joinmastodon.org/2018/06/how-to-implement-a-basic-activitypub-server/)) (#415)
  - [ ] Use webfinger to detect Lemmy instances, before calling Lemmy specific APIs
  - [ ] Enforce https for federation when running in production mode
- [ ] Moderation
  - [ ] Admin page to whitelist instances for federation
  - [ ] Admin/user options to mute/block remote users and instances
  - [ ] Reporting across instances
- [ ] Testing
  - [x] build a test environment ([!7](https://yerbamate.dev/dessalines/lemmy/pulls/7))
  - [ ] integration suite for testing (#147)

[This document](https://github.com/dessalines/lemmy/blob/master/docs/apub_api_outline.md) has an outline with the ActivityPub objects that we will need (but its far from the complete picture).

**Comment by nelsonblaha** [comment]
Eagerly awaiting this so I can stand up my Lemmy and join the fediverse!

**Comment by dessalines** [comment]
It won't be for a while yet, still a lot of dev work to do. Also you won't see users in your feed / main page, you'd see posts from federated communities. lemmy is more of a reddit clone, and about communities, rather than a twitter clone like mastodon and pleroma.

**Comment by LeoPCI** [comment]
Is there anything people can do to contribute to federation?

**Comment by dessalines** [comment]
Ya, read through the codebase and [this activitypub outline](https://github.com/dessalines/lemmy/blob/master/docs/apub_api_outline.md), and start adding the vocabulary and endpoints.

**Comment by poVoq** [comment]
How will communities work?

Maybe a !bang tag compatible group channel solution like found in Hubzilla and Friendica would be a good option?

**Comment by dessalines** [comment]
You mean for autocomplete? Just use `#`.

**Comment by poVoq** [comment]
Ah, so communities will not be a separate channel to follow, but rather just filtered by # tags? That seems not ideal to be honest. There is an established methode (from the GNU Social days) that in implemented in Friendica and Hubzilla (and an PR exists for Pleroma) to manage it with !bang syntax. 

See also this related discussion about adding it to Mastodon:
https://github.com/tootsuite/mastodon/issues/139

**Comment by Nutomic** [comment]
@dessalines Here is a [Rust library for ActivityPub](https://crates.io/crates/activitypub), are you aware of that?

Also, I'm wondering what would be a good first step for implementing activitypub. Maybe implementing the actor for communities, and allowing other activitypub software to follow lemmy communities?

**Comment by dessalines** [comment]
Ya I'm using it. The hard thing I'm running into now is trying to get every piece of information jammed into the activitypub vocab. 

[Here are the 2 actors I picked out](https://github.com/dessalines/lemmy/blob/master/docs/apub_api_outline.md#actors), `Person` IE userand `Group` IE community. 

At first I was gonna get federated community follows working, but then at some point allow subscribing to users, and having user posts show up in your feed, just like mastodon.

**Comment by dessalines** [comment]
Here's the one I have started on for person: https://github.com/dessalines/lemmy/blob/master/server/src/apub.rs

**Comment by alexbuzzbee** [comment]
Federation needs to go to the top of the priority list if it isn't already there. It's the one big blocker standing between Lemmy and actual use in more than a couple of places.

**Comment by Nutomic** [comment]
Updated the initial post with a more complete list.

**Comment by Nutomic** [comment]
Updated it again, with more details. As you can see, there is a ton of work to do.

Also check out my [PR here](https://yerbamate.dev/dessalines/lemmy/pulls/7), which implements some very basic federation and a test setup.

**Comment by dessalines** [comment]
@Nutomic Thx for coming up with this outline btw. Imma start working on some of the community federation stuff within the next few days.

**Comment by StaticallyTypedRice** [comment]
> Federation needs to go to the top of the priority list if it isn't already there. It's the one big blocker standing between Lemmy and actual use in more than a couple of places.

I agree. There are so many nested comment forums or "Reddit alternatives" out there but in the end they ALL suffer the same issue of there being a single point of control, which is what drives many people away from Reddit and other social media in the first place. We absolutely need federation if we're going to have any hope of becoming a next-gen social media platform.

**Comment by Nutomic** [comment]
Closing in favour of #647. @dessalines if you think there is anything important here, please add it to the new issue.
